### PR TITLE
drivers: spi: esp32xx: Fix word size limiting transaction

### PR DESCRIPTION
**Pull request by dean-at-evos**
Word size setting limited any SPI transaction to the frame size. In addition to making the SPI inefficient this broke drivers that set the word size. It appears that most drivers use a one byte (8) size for this setting.
This change respects what I think is the intended use of the word size setting. That is to set the length of each element in a tx/rx buffer struct.

In my use case this change made the driver for the ENC28J60 SPI ethernet controller hang on initialisation.

### CODE REVIEW

**Review by LucasTambor — COMMENTED** [review]
Thanks for this PR. So, this raised a doubt in my mind about the function of word size. I though it meant to define the size of each transaction. You are saying that this is not true, but to organise the internal rx/tx buffer?

**Comment by sylvioalves** [comment]
@dean-at-evos, your PR message says `In my use case this change made the driver for the ENC28J60 SPI ethernet controller hang on initialisation.`. Do you mean current implementation causes the hang and this PR fixes it? It seems very similar to previous version after @LucasTambor update this to fix thw word size issue (8 bytes ~ 32 bytes len).

**Comment by dean-at-evos** [comment]
> @dean-at-evos, your PR message says `In my use case this change made the driver for the ENC28J60 SPI ethernet controller hang on initialisation.`. Do you mean current implementation causes the hang and this PR fixes it? It seems very similar to previous version after @LucasTambor update this to fix thw word size issue (8 bytes ~ 32 bytes len).

Apologies, my unclear writing. What I meant was that the commit @LucasTambor made previously,  #56283, broke that particular driver.
There is some confusion about what that setting is for as far as I can tell. The [documentation](https://docs.zephyrproject.org/latest/hardware/peripherals/spi.html#structspi__config) for it isn't clear and the way it is used in other drivers doesn't seem consistent...
In my experience most SPI peripherials expect a 'transaction' to be completed in one /CS cycle. That is, assert CS, send/recieve data, then deassert CS. Additionally a peripherial MAY have a minimum data unit frame size (most seem to be a byte).
So, in my interpretation of the intent here, the transaction length (that is the time that CS is asserted) is set by the length of the buffer set you wish to send. The units of length for the data inside the buffers is set by the WORD SIZE setting.
Of course I could be wrong here, but this is the approach that seems to work for the drivers I have used, and am writing, and the ones I have looked at to try and work out what is going on :)
Maybe we need to start a conversation somewhere more general to get a good idea of the intent of this setting?

File: drivers/spi/spi_esp32_spim.c [hunk 1]
```diff
@@ -354,11 +356,14 @@ static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
 static inline uint8_t spi_esp32_get_frame_size(const struct device *dev,
```
[the remark this answers is not in the document]
Comment (dean-at-evos) [reply drivers/spi/spi_esp32_spim.c:356]: Probably not....I'll fix

**Comment by LucasTambor** [comment]
@dean-at-evos , I understand your point and actually agree with you on how this setting should be used. The changes I've made (#56283) were motivated by this issue:   #54746. I understand at the time that this setting should limit every transaction (cs cycle) to WORD SIZE. Living up to the buffer size when WORD SIZE was not set. 

How can we get a definitive answer regarding this setting?

**Comment by dean-at-evos** [comment]
@

> @dean-at-evos , I understand your point and actually agree with you on how this setting should be used. The changes I've made (#56283) were motivated by this issue: #54746. I understand at the time that this setting should limit every transaction (cs cycle) to WORD SIZE. Living it to the buffer size when WORD SIZE was not set.
> 
> How can we get a definitive answer regarding this setting?

I did see #54746 and I think that there was an issue with the setting for ESP32 driver in that it didn't treat the word size/frame size correctly. If you look at the example for the issue they have set the word size to 32 bits in the buffer and then the length to 4, it should have been 1, but it still may not have worked as intended. This fragment from ```spi_context.h``` shows that the buffer ```len``` property is treated as word length, not byte length. The dfs (data frame size, set from the WORD SIZE config) is used with the ```buf.len``` to get the byte len.
```c
	ctx->rx_len -= len;
	if (!ctx->rx_len) {
		/* Current buffer is done. Get the next one to be processed. */
		++ctx->current_rx;
		--ctx->rx_count;
		ctx->rx_buf = (uint8_t *)
			spi_context_get_next_buf(&ctx->current_rx,
						 &ctx->rx_count,
						 &ctx->rx_len, dfs);
	} else if (ctx->rx_buf) {
		ctx->rx_buf += dfs * len;
	}
```
Also examining ```spi_bitbang.c``` gives some clues as to how the device is expected to function.
I think that with this change the ESP spi controller driver now works as bus peripheral drivers expect, and also issue #54746 is fixed (as well as the heap corruption bug, which was a real nasty one).
It would be nice if the documentation was much clearer around the intent of these settings, and what driver writers/implementers should expect....but like all of us documentation comes last :)

**Review by LucasTambor — APPROVED** [review]

**Review by sylvioalves — APPROVED** [review]

**Pull request merged**